### ISSUE TITLE
C++ Tensor.clone() changes max threads in OpenMP

### ISSUE THREAD

**Issue by uurieka** [pytorch/pytorch#117930]
### 🐛 Describe the bug

When trying to clone a `torch::Tensor` with many rows it changes the max number of threads in OMP:

```
torch::Tensor t1 = torch::empty({32950,300});
cout << "A. Max threads in omp: " << omp_get_max_threads() << endl;
torch::Tensor t2 = t1.clone();
cout << "B. Max threads in omp: " << omp_get_max_threads() << endl;
```

Result:

```
A. Max threads in omp: 8
B. Max threads in omp: 4
```

However, if the tensor is "small" then it does not happen:

```
torch::Tensor t1 = torch::empty({20,20});
cout << "A. Max threads in omp: " << omp_get_max_threads() << endl;
torch::Tensor t2 = t1.clone();
cout << "B. Max threads in omp: " << omp_get_max_threads() << endl;
```

Result:

```
A. Max threads in omp: 8
B. Max threads in omp: 8
```

Same problem happens when building a data_loader `torch::data::make_data_loader(ds, options)`

A temporary solution that worked for me:

```
int th = omp_get_max_threads();
torch::Tensor t1 = torch::empty({32950,300});
cout << "A. Max threads in omp: " << omp_get_max_threads() << endl;
torch::Tensor t2 = t1.clone();
cout << "B. Max threads in omp: " << omp_get_max_threads() << endl;
omp_set_num_threads(th);
cout << "C. Max threads in omp: " << omp_get_max_threads() << endl;
```

Result:

```
A. Max threads in omp: 8
B. Max threads in omp: 4
C. Max threads in omp: 8
```

The problem is that libtorch is deliberately changing the max threads in OpenMP, in this case the developer does not notice the change and the program continues working at reduced capacity.

### Versions

Collecting environment information...
PyTorch version: 2.1.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.2.0-39-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1660 SUPER
Nvidia driver version: 535.104.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             8
On-line CPU(s) list:                0-7
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 5 3400G with Radeon Vega Graphics
CPU family:                         23
Model:                              24
Thread(s) per core:                 2
Core(s) per socket:                 4
Socket(s):                          1
Stepping:                           1
Frequency boost:                    enabled
CPU max MHz:                        3700.0000
CPU min MHz:                        1400.0000
BogoMIPS:                           7386.42
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          128 KiB (4 instances)
L1i cache:                          256 KiB (4 instances)
L2 cache:                           2 MiB (4 instances)
L3 cache:                           4 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-7
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT vulnerable
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] onnxruntime==1.16.3
[pip3] pytorch-lightning==1.9.4
[pip3] torch==2.1.2
[pip3] torchmetrics==1.2.1
[pip3] torchsde==0.2.5
[pip3] torchvision==0.16.2
[pip3] triton==2.1.0
[conda] Could not collect

**Comment by tringwald** [comment]
I can reproduce this with latest libtorch. For small tensors, the number of threads is not changed, while for large tensors the number of threads is decreased after the `.clone()` call.

However, in both cases setting `torch::set_num_threads()` is respected and the number of threads remains constant.

**Comment by uurieka** [comment]
Thanks for the reply

The method `torch::set_num_threads(int)` works as explained (it should be called at the beginning of the program to avoid the problem). 

On the other hand, I found that even the`torch::get_num_threads()` method affects OMP:

```
cout << "Max OMP  : " << omp_get_max_threads() << endl;
cout << "Max Torch: " << torch::get_num_threads() << endl;
cout << "Max OMP  : " << omp_get_max_threads() << endl;
```

Result:

```
Max OMP  : 8
Max Torch: 4
Max OMP  : 4
```

I would describe this as an intrusive behaviour. But still the `torch::set_num_threads(int)` method is a temporary solution to avoid the problem.

**Comment by tringwald** [comment]
I agree that the behavior is a bit confusing. However, in both examples, `torch` also respects the `OMP_NUM_THREADS` env variable and does not change the thread count any further.

What we might want to change is that `torch` seems to default to `cpuinfo_get_processors_count()` (16 physical cores for me) while OMP seems to default to the logical core count including SMT (32 for me).

**Comment by mingfeima** [comment]
@sanchitintel could you please help debug this one?

**Comment by jgong5** [comment]
> What we might want to change is that `torch` seems to default to `cpuinfo_get_processors_count()` (16 physical cores for me) while OMP seems to default to the logical core count including SMT (32 for me).

The reason why only physical cores are used is to avoid resource contention shared among SMTs in a single physical core.

**Comment by uurieka** [comment]
> > What we might want to change is that `torch` seems to default to `cpuinfo_get_processors_count()` (16 physical cores for me) while OMP seems to default to the logical core count including SMT (32 for me).
> 
> The reason why only physical cores are used is to avoid resource contention shared among SMTs in a single physical core.

If this is a restriction/control, how dangerous can be this at the beginning of the program?

```
int main(){
    torch::set_num_threads( omp_get_max_threads() );
    // ...
}
```

Should we consider other options to avoid `torch` overriding/overwriting the user/dev custom/standard OMP config? Ex:

1. `torch` to restore the OMP max threads each time that `torch` affects it for internal purposes.
2. Document the behaviour so the user/dev can manually restore the OMP max threads each time that `torch` affects it.
3. To isolate the max threads in `torch` so OMP is not affected.

**Comment by sanchitintel** [comment]
Sorry, I only noticed today that this issue had been assigned to me. 😞 

The number of OpenMP threads are lazily set to the number of physical cores when the first parallel call is made, or when `get_num_threads` is called.
@uurieka, when you used a small tensor, using multiple threads were not required for that computation.
When you used a larger tensor with `clone`, it was considered the first parallel call.

I see your point that documenting this behavior would help libtorch users (developers who use libtorch in their projects).
The other options that you suggested do not seem feasible.

**Comment by sanchitintel** [comment]
> If this is a restriction/control, how dangerous can be this at the beginning of the program?
> 
> ```
> int main(){
>     torch::set_num_threads( omp_get_max_threads() );
>     // ...
> }
> ```
> 

You can profile your workload, and if necessary, change the size of the OpenMP thread-pool later as well, if you'd notice that using more threads than the number of physical cores is hindering performance.
